client/allocrunner/taskrunner/api_hook.go

```diff
@@ -0,0 +1,113 @@
+package taskrunner
+
+import (
+	"context"
+	"errors"
+	"net"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/client/allocdir"
+	"github.com/hashicorp/nomad/client/allocrunner/interfaces"
+	"github.com/hashicorp/nomad/client/config"
+)
+
+// apiHook exposes the Task API. The Task API allows task's to access the Nomad
+// HTTP API without having to discover and connect to an agent's address.
+// Instead a unix socket is provided in a standard location. To prevent access
+// by untrusted workloads the Task API always requires authentication even when
+// ACLs are disabled.
+//
+// The Task API hook largely soft-fails as there are a number of ways creating
+// the unix socket could fail (the most common one being path length
+// restrictions), and it is assumed most tasks won't require access to the Task
+// API anyway. Tasks that do require access are expected to crash and get
+// rescheduled should they land on a client who Task API hook soft-fails.
+type apiHook struct {
+	shutdownCtx context.Context
+	srv         config.APIListenerRegistrar
+	logger      hclog.Logger
+	ln          net.Listener
+}
+
+func newAPIHook(shutdownCtx context.Context, srv config.APIListenerRegistrar, logger hclog.Logger) *apiHook {
+	h := &apiHook{
+		shutdownCtx: shutdownCtx,
+		srv:         srv,
+	}
+	h.logger = logger.Named(h.Name())
+	return h
+}
+
+func (*apiHook) Name() string {
+	return "api"
+}
+
+func (h *apiHook) Prestart(_ context.Context, req *interfaces.TaskPrestartRequest, resp *interfaces.TaskPrestartResponse) error {
+	if h.ln != nil {
+		// Listener already set. Task is probably restarting.
+		return nil
+	}
+
+	udsPath := apiSocketPath(req.TaskDir)
+	if err := os.RemoveAll(udsPath); err != nil {
+		h.logger.Warn("error removing task api socket", "path", udsPath, "error", err)
+	}
+
+	udsln, err := net.Listen("unix", udsPath)
+	if err != nil {
+		h.logger.Warn("error listening on task api socket", "path", udsPath, "error", err)
+		return nil
+	}
+
+	if err := os.Chmod(udsPath, 0o777); err != nil {
```

Comment:
I think our logic here for why this is safe in the context of the task dir is sound, but it's also going to result in us getting dinged by third-party auditors armed with sophisticated source code analysis tools like `grep` :grinning: 

Can we leave a comment above here about why it's ok in hopes of reducing the amount of noise we get from those discussions?

Reply:
Good points. I was going to say `well we require auth anyway so who cares` but...

These are the sorts of decisions we have to live with forever, and respecting `Task.User` when its set (as `nomad_token` does too) seems like a valuable road to go down.

So I followed #15755 and made this attempt to chown the socket to `Task.User` with `0o600` perms. Like #15755 it falls back to `0o666` if that fails (which it always will on Windows or as a non-root user).